design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+
+This KLIP proposes a redesign of the query language but does not propose to implement all possible features right away.
+The intent is to make the language "future proof" such that new feature can be added seamlessly without breaking the language.
+In particular, we should not implement (cf. "Design" section below):
+
+* re-introduce `CREATE TABLE` statement to create RDBMS style tables that allow insert/update/delete statements
+* allow for complex queries over table state, e.g., aggregation and joins
+* we might not need to implement the new `AS OF <timestamp>` clause in the first iteration
+* we might not need to support terminating queries over streams right away
+
+We limit this KLIP to not introduce too much syntactic sugar to express queries.
+This may make some queries unnecessary "complex/clumsy" but the goal is to illustrate to underlying principles of the language.
+We can add syntactic sugar after we agree on the basic design of the language.
+
+
+
+## Value/Return
+
+The "Dual Query Language" should be easier to understand and easier to use.
+It will allow to support nested queries seamlessly and unifies query semantics (persistent vs. transient queries).
+Furthermore, the language is designed to be "future proof", i.e., it considers potential future features and thus allows
+to extend the capabilities of KSQL without the need to change the language in a breaking way.
+
+
+
+## Public APIS
+
+We propose the following language changes:
+
+* introduce `TABLE(<stream>)` operator: takes a STREAM as input, and changes its semantic interpretation from facts to updates; it returns a TABLE
+  * note: the returned TABLE is not an RDBMS style table, but rather a MV
+  * note: it's not specified if the returned TABLE is materialized into a state store or represented as a changelog
+    (it's up the the KSQL optimizer or KS runtime to make a materialization decision)
+  * the `TABLE()` operator consumes it's input STREAM "from beginning"
+  * we could extend `TABLE()` to also accept a TABLE as input if it makes the language smoother;
+    it's obvious that `TABLE(<table>)` would be idempotent, i.e., `TABLE(t) = t`
+* introduce `STREAM(<table>)` operator: takes a TABLE as input, and returns a STREAM that contains a record for each update to the TABLE
+  * note: the returned STREAM is a fact stream, and would be processed with corresponding STREAM semantics (it's semantically not a changelog stream)
+  * note: the input TABLE can either be an RDBMS style table of a MV
+  * by default, the output STREAM contains one record for each input table record (full table scan on the table generation when the query was issues)
+    plus a record for each input table update
+  * we could extend `STREAM()` to also accept a STREAM as input if it makes the language smoother;
+  it's obvious that `STREAM(<stream>)` would be idempotent, i.e., `STREAM(s) = s`
+* Introduce an `AS OF <timestamp>` clause for RDBMS style queries against tables
```

Comment:
I have the impression the KLIP does a bad job explaining what it proposes... Can you clarify your comment?

> This seems like an odd knob to twist to change from streaming to RDBMS style queries... 

`AS OF <timestamp>` has nothing to do with that -- it allows you to access an older version of a table (something that is not supported atm in ksqlDB at all). Maybe I am misunderstand what you are saying?

> If you want to query tables in a streaming fashion (i.e. subscribe to changes) then you should add EMIT CHANGES or EMIT FINAL.

Well, this KLIP proposes to use `STREAM(<table>)` to do this.

> I still don't think there's a need for RDBMS style queries against streams yet

Agreed, and I this KLIP does not propose it. Not even sure what an "RDBMS style queries against streams" actually is? Can you elaborate?

Reply:
I totally misunderstood what this was until I read the examples later on. You meant to do _temporal_ queries on tables. I had assume you meant "this makes it a traditional RDBMS when otherwise it wouldn't be".

I think it all makes sense now - you can ignore this comment.